### ISSUE TITLE
Full analog support

### ISSUE THREAD

**Issue by Fortinbra** [OpenStickCommunity/GP2040-CE#294]
Placeholder for completing the analog stick/trigger support.

- [ ] Support multiples of the ADC chip for additional channels
- [ ] Support mixing onboard ADC with external ADC for inputs
- [x] Support LS/RS as true analog stick
- [ ] Support LT/RT as true analog triggers
- [ ] Implement calibration and web interface calibrating mapped analog inputs
- [ ] (Optional) Add automatic calibration option

**Comment by TheTrainGoes** [comment]
We want to look into something before doing this - It is possible that dual analog is possible through a broken out embedded RP2040.  

We would like to add an action item to this to expand the current analog offering to take advantage of broken out RP2040 chips to see if things can be handled as is.

**Comment by TheTrainGoes** [comment]
We have a working PoC for dual analog support now via a broken out RP2040 chip that has access to all of the ADC pins.

**Comment by mordae** [comment]
Hi. I've just accidentally stumbled upon this project and wanted to chime in.

It is possible to use any RP2040 GPIO pin as ΣΔ ADC. To get an idea, check out [this simulation](https://tinyurl.com/2ecqbbp7) I've put together.

As for the implementation, the simplest way would be to:

- Configure the GPIO for output
- Enable pull-down
- Emit 1 (to quickly charge the capacitor)
- Configure GPIO for input (to slowly drain the capacitor)
- Read GPIO until it switches from 1 to 0, incrementing a counter
- Counter value is a usable readout

You can experiment with the passives to get a feel for what would work best.

The ideal implementation would be a little more involved. I have used PIO to get best results. In a nutshell:

- Configure the GPIO for input and use the "bus keep" mode by enabling both pull-up and pull-down. Now when you output something on the GPIO, you are actually controlling the pull-up/down and thus have about ~50kΩ series resistance. This is going to smoothly steer the capacitor voltage.
- Enable PIO for the pin and write a (possibly single instruction) program that outputs negation of the pin back to itself. Since we are in the bus keep mode and there is a relatively large capacitor, we have just constructed a (rather gentle) oscillator.
- Write another PIO program that continuously decrements one of the registers when the pin is high and the other when it is low using branch on pin value. To take measurements:
  - Inject instructions to reset the registers to 0.
  - Let it run for a while, possibly doing other stuff such as checking buttons and so on. Ideally it should run for 10 ms so that you get 100 Hz output, but it's really up to you.
  - Inject push X, push Y to get the counters to the FIFO and then read them from there. Subtract them from 0 to get positive numbers. Then subtract one from the other to get your readout.

You need to calibrate every stick before it gets usable. Depending on capacitor and joystick tolerances there will be differences between readouts of different axes, so you really need to calibrate all of them to get minimum, maximum and center points.

This mechanism is similar to how joysticks were read back in the [Game port](https://en.wikipedia.org/wiki/Game_port) days.

**Comment by TheTrainGoes** [comment]
We have now rolled out dual analog for RP2040 based devices that have access to all four ADC pins.

**Comment by osdx190** [comment]
**Hi, I've tried your latest update and I've had a problem with the analog ones not giving the maximum values** and it would be because they are 60° resistors on the joystick. I have mapped it with windows and it works but I use it in a game or program it doesn't work at all there are sides that give me like 93% movement, one side gives me 70% and others 81%. these are my external joystick calibration ideas. Although it occurred to me to add capacitors as in the original boards.
![external ring calibration](https://github.com/OpenStickCommunity/GP2040-CE/assets/140558392/5281d3b8-ab6c-4d3e-bd8f-8d031f7d1438)
**and i have managed to add an adc on the standard raspberry pi pico board.**

![add ADC_3 to raspberry pi pico](https://github.com/OpenStickCommunity/GP2040-CE/assets/140558392/e99a2d7a-4512-4644-bc65-e1582cac6e02)
